docs(fix): Improve `<source>` element page for consistency and clarity (#29946)
mdn/content: files/en-us/web/html/element/source/index.md

{{HTMLSidebar}}

> The **`<source>`** [HTML](/en-US/docs/Web/HTML) element specifies multiple media resources for the {{HTMLElement("picture")}}, {{HTMLElement("audio")}}, and {{HTMLElement("video")}} elements. It is a {{glossary("void element")}}, which means that it has no content and does not require a closing tag. This element is commonly used to offer the same media content in multiple file formats in order to provide compatibility with a broad range of browsers given their differing support for [image file formats](/en-US/docs/Web/Media/Formats/Image_types) and [media file formats](/en-US/docs/Web/Media/Formats).

Reviewer: "multiple media resources" implies that you have to include more than one, whereas I'm pretty sure you can include just one. "one or more media resources"?

## Attributes

This element supports all [global attributes](/en-US/docs/Web/HTML/Global_attributes). In addition, the following attributes can be used with it:

- `type`

  - : Specifies the [MIME media type of the image](/en-US/docs/Web/Media/Formats/Image_types) or [other media type](/en-US/docs/Web/Media/Formats/Containers), optionally including a [`codecs` parameter](/en-US/docs/Web/Media/Formats/codecs_parameter).

- `src`

  - : Specifies the URL of the media resource. Required if the parent of `<source>` is {{HTMLElement("audio")}} or {{HTMLElement("video")}}. Not allowed if the parent is {{HTMLElement("picture")}}.

- `srcset`

  - : Specifies a comma-separated list of one or more image URLs and their descriptors. Required if the parent of `<source>` is {{HTMLElement("picture")}}. Not allowed if the parent is {{HTMLElement("audio")}} or {{HTMLElement("video")}}.

    The list consists of strings separated by commas, indicating a set of possible images for the browser to use. Each string is composed of:

>     - A URL for an image.

Reviewer: ```suggestion
    - A URL specifying an image location.
```

>     - An optional width descriptor—a positive integer directly followed by `"w"`, such as `300w`. If not specified, the value used by default is infinity.

Reviewer: ```suggestion
    - An optional width descriptor—a positive integer directly followed by `"w"`, such as `300w`. If not specified, the value used by default is `infinity`.
```

Author: This suggestion led me to check the spec (should have done earlier). I did not find a reference either in the [source element](https://html.spec.whatwg.org/multipage/embedded-content.html#the-source-element) page or in the linked [srcset attribute](https://html.spec.whatwg.org/multipage/images.html#srcset-attributes) page.

This section in the spec from "srcset attribute" does not mention any default values for the width and pixel density descriptors. I am not sure if we should mention them in our docs. What do you think?

> 4. Zero or one of the following:
> 
>     - A width descriptor, consisting of: [ASCII whitespace](https://infra.spec.whatwg.org/#ascii-whitespace), a [valid non-negative integer](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-non-negative-integer) giving a number greater than zero representing the width descriptor value, and a U+0077 LATIN SMALL LETTER W character.
> 
>     - A pixel density descriptor, consisting of: [ASCII whitespace](https://infra.spec.whatwg.org/#ascii-whitespace), a [valid floating-point number](https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#valid-floating-point-number) giving a number greater than zero representing the pixel density descriptor value, and a U+0078 LATIN SMALL LETTER X character.

Reviewer: Hrm, this is a good point. The default value descriptions aren't that helpful and probably confuse more than they inform. About the closest there is to a default value description in the spec is on the `srcset` page:

> For the purpose of this requirement, an image candidate string with no descriptors is equivalent to an image candidate string with a 1x descriptor.

A bigger issue, in my mind, is that the text here doesn't really describe the purpose of `srcset`/`sizes`, or how you would use them. 

I would suggest:

1. Removing the default value descriptions, or perhaps just say something afterwards to the effect that, if a `...w` or `...x` is not included, it is equivalent to specifying `1x`. I think that is what the spec is saying?
2. Include some more description and usage examples on the page, or link to some somewhere else. I did attempt to provide this in my learning area [Responsive Images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#how_do_you_create_responsive_images) article, but I don't know how bulletproof that actually is. It seems to make sense to me, but we've had quite a few tickets filed in the past querying whether the code is correct. It was challenging to figure this stuff out.

Author: Thanks for taking the time and for offering suggestions!
1. Agree, removed the default value statement from each descriptor and added a common one in the subsequent para (https://github.com/mdn/content/pull/29946/commits/c96683a2237958df96cbaa36e64da26cd96ebd6a).
2. Should we tackle this outside this PR? :)

- `media`

  - : Specifies the [media query](/en-US/docs/Web/CSS/CSS_media_queries) for the resource's intended media.

- `height`

  - : Specifies the intrinsic height of the image in pixels. Allowed if the parent of `<source>` is a {{HTMLElement("picture")}}. Not allowed if the parent is {{HTMLElement("audio")}} or {{HTMLElement("video")}}.

    The height value must be an integer without any units.

- `width`

  - : Specifies the intrinsic width of the image in pixels. Allowed if the parent of `<source>` is a {{HTMLElement("picture")}}. Not allowed if the parent is {{HTMLElement("audio")}} or {{HTMLElement("video")}}.

    The width value must be an integer without any units.

## Usage notes

The `<source>` element is a **{{glossary("void element")}}**, which means that it not only has no content but also has no closing tag. That is, you _never_ use "`</source>`" in your HTML.

> If the `type` attribute isn't specified, the browser retrieves the media's type from the server and determine if it can be displayed. If the media can't be rendered, the browser checks the next `<source>` in the list. If the `type` attribute is specified, the browser immediately compares it with the media types it can display. If the type is not supported, the browser skips querying the server and directly checks the next `<source>` element.

Reviewer: ```suggestion
The browser moves through a list of `<source>` elements to find a format it can understand. It will use the first one it can display. For each `<source>` element:

- If the `type` attribute isn't specified, the browser retrieves the media's type from the server and determines if it can be displayed. If the media can't be rendered, the browser checks the next `<source>` in the list.
- If the `type` attribute is specified, the browser immediately compares it with the media types it can display. If the type is not supported, the browser skips querying the server and directly checks the next `<source>` element.
```

Author: Ooh, this is much better, thanks! And I like it when we can break up text into points to aid in comprehension. I've mostly fixed as suggested with just tiny tweaks to "moves" and "understand":

> The browser goes through a list of `<source>` elements to find a format it supports. It uses the first one it can display. For each `<source>` element:
> 
> - If the `type` attribute isn't specified, the browser retrieves the media's type from the server and determines if it can be displayed. If the media can't be rendered, the browser checks the next `<source>` in the list.
> - If the `type` attribute is specified, the browser immediately compares it with the media types it can display. If the type is not supported, the browser skips querying the server and directly checks the next `<source>` element.

> When the `<source>` elements are used within a `<picture>` element, the browser will fall back to using the image specified in the `<picture>` element's {{HTMLElement("img")}} child. This happens if none of the `<source>` elements provide a suitable image.

Reviewer: ```suggestion
If none of the `<source>` elements provide a usable source:

- In the case of a `<picture>` element, the browser will fall back to using the image specified in the `<picture>` element's {{HTMLElement("img")}} child.
- In the case of an `<audio>` or `<video>` element, the browser will display the fallback content included between the element's opening and closing tags. 
```

Check this last fact; I'm pretty sure this is the case...

Author: Seems correct. We say so in our [`<audio>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/audio) and [`<video>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video) docs:

> The content inside the opening and closing `<audio></audio>` tags is shown as a fallback in browsers that don't support the element.

For information about image formats supported by web browsers and guidance on selecting appropriate formats to use, see our [Image file type and format guide](/en-US/docs/Web/Media/Formats/Image_types). For details on the video and audio media types you can use, see the [Media type and format guide](/en-US/docs/Web/Media/Formats).

## Examples

> ### Using type attribute with video

Reviewer: ```suggestion
### Using the type attribute with `<video>`
```

Author: This led me to go check our [decision](https://github.com/orgs/mdn/discussions/271#discussioncomment-4225841) about using code styling in headings 🙂.

I've updated as suggested. Also added backticks around the attribute names in all example headings

</video>
```

> ### Using media attribute with video

Reviewer: ```suggestion
### Using the media attribute with `<video>`
```

> This example demonstrates how to offer an alternate source file for viewports that are at or above a certain width. When a user's browsing environment meets the specified `media` condition, the associated `<source>` element is chosen. The contents of its `src` attribute are then requested and rendered. If the `media` condition does not match, the browser will move on to the next `<source>` in the list. The second `<source>` option in the code below has no `media` condition, so it will be selected for all other browsing contexts.

Reviewer: ```suggestion
This example demonstrates how to offer an alternate source file for viewports above a certain width. When a user's browsing environment meets the specified `media` condition, the associated `<source>` element is chosen. The contents of its `src` attribute are then requested and rendered. If the `media` condition does not match, the browser will move on to the next `<source>` in the list. The second `<source>` option in the code below has no `media` condition, so it will be selected for all other browsing contexts.
```

</video>
```

For more examples, the [Video and audio content](/en-US/docs/Learn/HTML/Multimedia_and_embedding/Video_and_audio_content) article in the Learn area is a great resource.

> ### Using media attribute with picture

Reviewer: ```suggestion
### Using the media attribute with `<picture>`
```

</picture>
```

With the `<picture>` element, you must always include an `<img>` with a fallback image. Also, make sure to add an `alt` attribute for accessibility, unless the image is purely decorative and irrelevant to the content.

> ### Setting height and width attributes in picture

Reviewer: ```suggestion
### Setting height and width attributes in `<picture>`
```

For consistency with my previous suggestions ;-) ?

Author: I've updated this to:

> "### Using `height` and `width` attributes with `<picture>`

...because the attributes are set in `<source>`

Reviewer: I'm assuming you decided against this in the end because it made the heading overly long. I didn't see this in the updated preview file. Either is OK by me.

Author: Oh! I see the updated heading in the [commit](https://github.com/mdn/content/pull/29946/commits/fb77f78934aa5cccf93b48366798337e4055f272#diff-4159a52f149fe523ff90e751b0fbee114521fd997d6724e5f26e7fc36e8abf17R173) and [preview page](https://pr29946.content.dev.mdn.mozit.cloud/en-US/docs/Web/HTML/Element/source#using_height_and_width_attributes_with_picture).
